bufferserver/src/main/java/com/datatorrent/bufferserver/internal/DataList.java

```diff
@@ -405,7 +405,27 @@ public boolean resumeSuspendedClients(final int numberOfInMemBlockPermits)
 
   public boolean isMemoryBlockAvailable()
   {
-    return (storage == null) || (numberOfInMemBlockPermits.get() > 0);
+    return (numberOfInMemBlockPermits.get() > 0) || ((storage == null) && !areSubscribersBehindByMax());
```

Comment:
Move check `storage == null` inside areSubscribersBehindByMax(). Can you pass a number of blocks that the publisher is ahead of subscribers as a parameter? Currently, it reuses max number of blocks, that in general should include all allocated blocks. In future, back pressure should provide a separate parameter that tells by how many blocks publisher can be ahead. Please also consider renaming areSubscribersBehind to isPublisherAhead.